Redesign Share UI to emphasize difference between Original URLs and Snapshot URLs.

cjcenizal opened the pull request:
This supersedes https://github.com/elastic/kibana/pull/8157.

Redesign Share UI to emphasize difference between Original URLs and Snapshot URLs.
- Addresses https://github.com/elastic/kibana/issues/8051
- Blocked by https://github.com/elastic/kibana/issues/8022
## Discover
### Unsaved

![image](https://cloud.githubusercontent.com/assets/1238659/18217111/dabb4bf6-710f-11e6-8d93-efe518069fb9.png)
### Saved

![image](https://cloud.githubusercontent.com/assets/1238659/18217116/df9e18a6-710f-11e6-802a-457d0346699a.png)
### Short URL

![image](https://cloud.githubusercontent.com/assets/1238659/18217127/e5a2ea7e-710f-11e6-90a8-edeeb5cdd93a.png)
## Dashboard
### Unsaved

![image](https://cloud.githubusercontent.com/assets/1238659/18217139/ee68a054-710f-11e6-8831-7abbdbdf2521.png)
### Saved

![image](https://cloud.githubusercontent.com/assets/1238659/18217150/f5efb204-710f-11e6-93bb-3c653eea60e3.png)
### Short URL

![image](https://cloud.githubusercontent.com/assets/1238659/18217156/fbb99074-710f-11e6-8676-0c4942164204.png)
## Visualization
### Unsaved

![image](https://cloud.githubusercontent.com/assets/1238659/18217161/07ecc780-7110-11e6-8f90-dfeeec298437.png)
### Saved

![image](https://cloud.githubusercontent.com/assets/1238659/18217163/0cee52f8-7110-11e6-8c1e-ea4ae92add4e.png)
### Short URL

![image](https://cloud.githubusercontent.com/assets/1238659/18217166/1244fd38-7110-11e6-8d0e-14d7dca204e0.png)

cjcenizal commented:
I got some design feedback from @alt74 and we updated the UI to look like this:

![image](https://cloud.githubusercontent.com/assets/1238659/18286867/8e653236-7429-11e6-81fc-ece13efb16d6.png)

spalger commented:
For the record, I thought the yellow warning text 💯 . It's seems pretty easy to miss completely now.

spalger commented on src/ui/public/share/views/share.html:
```diff
@@ -1,15 +1,140 @@
+
+      <!-- Notes -->
+      <div class="share-panel-form-note" ng-if="!share.shortSnapshotUrl">
+        We recommend IE users share a Short URL due to IE’s restrictions on URL length.
```
I think this should be more about compatibility with consumers of the URL, rather than sharing it. Maybe something like...

```
We recommend sharing shortened snapshot urls for maximum compatibility. Internet Explorer has URL length restrictions, and some wiki and markup parsers don't do well with the full-length version of the snapshot URL, but the short URL should work great.
```

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,109 @@
 const app = uiModules.get('kibana');
 
-app.directive('share', function () {
+app.directive('share', function (Private, Notifier) {
```
Would you mind importing Notifier from `ui/notify/notifier` instead?

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,109 @@
       objectType: '@',
       objectId: '@',
-      setAllowEmbed: '&?allowEmbed'
+      allowEmbed: '@allowEmbed'
```
mentioning the attribute name is redundant now.

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,109 @@
+    controllerAs: 'share',
+    controller: function ($scope, $location, globalState) {
+      // Default to allowing an embedded IFRAME, unless it's explicitly set to false.
+      this.allowEmbed = $scope.allowEmbed === 'false' ? false : true;
```
I kind of prefer the expression/attribute-value-as-a-function here. Mind explaining why you went for string comparison instead?

spalger replied:
I'm coming to terms with the fact that `@` is a bit more predictable in this scenario, especially if we aren't going to bind to these values over time. Do you think it would make sense to validate this scope value a bit though?

cjcenizal replied:
What do you mean by validating the attribute value? Check if it's either 'true' or 'false'?

spalger replied:
Yeah, it's maybe it's not a big deal, but anytime there is a whitelist of string values allowed I think it is important to validate. Especially when `"fals"` or `"fasle"` will be treated as `true`

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,109 @@
+
+      function getOriginalUrl(url) {
+        // If there is no objectId, then it isn't saved, so it has no original URL.
+        if ($scope.objectId === undefined || $scope.objectId === '') {
```
If this function is going to reach to `$scope` for the `objectId`, maybe it could also reach out for the `url`? It seems strange that one is an arg but not both/neither.

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,109 @@
+      this.getEmbeddableUrl = url => {
+        const embedQueryParam = '?embed=true';
+        const urlHasQueryString = url.indexOf('?') !== -1;
+        const embeddableUrl =
```
This would probably read better with an `if (x) { return y }` vs. the multi-line ternary

spalger commented:
If you get a short url, and then the url changes, the short url persists and points to the wrong long-url.

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,109 @@
+      };
+
+      // Original URLs.
+      this.originalUrl;
```
How would you feel about putting all the urls under `this.urls = { original, snapshot, ... }`. Then, in `updateUrls()` you can just overwrite the `this.urls` object in order to deal with https://github.com/elastic/kibana/pull/8172#issuecomment-245335433

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,109 @@
+        this.snapshotUrl = getSnapshotUrl(url);
+      };
+
+      updateUrls();
```
This shouldn't be necessary, the `$watch` will fire on initialization.

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,109 @@
+      $scope.$watch($location.absUrl, updateUrls);
+
+      this.useShortUrl = (shortUrlProperty, url) => {
+        if (this[shortUrlProperty]) return;
```
I think I would prefer different methods for each url type (ie. `this.enableShortSnapshotIframeUrl()`) rather than accepting the property name of the url here. I also think we need to be cautious of the race condition here

cjcenizal replied:
This sounds like a good idea, but what's the race condition you're thinking of? Someone clicking the "Shorten URL" link multiple times while the first request is still being processed on the back end?

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,109 @@
+        try {
+          // Select the text to be copied. If the copy fails, the user can easily copy it manually.
+          const copyTextarea = document.querySelector(selector);
+          copyTextarea.select();
```
Since the catch clause assumes that this line succeeds every time, perhaps it should be outside of the `try/catch`?

spalger commented on src/ui/public/share/styles/index.less:
```diff
@@ -1,21 +1,79 @@
+  margin-left: 30px;
+}
+
+  .share-panel-title {
```
Indentation? I'm sure this is intentional, but I'm curious why this isn't `.share-panel--title`

cjcenizal replied:
Good catch, I will change the name, which will justify the indentation (to indicate the parent-child relationship visually).

spalger commented:
Some notes, only one bug, looking good

cjcenizal commented:
Re the warning, I think it's better to present it subdued, since we don't really want to draw the user's attention to it if they're looking elsewhere. If the user is trying to share the Original, they'll be looking in that area already so I think the chances are good that they'll read the warning, despite the subdued presentation.

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,145 @@
+          snapshot: getSnapshotUrl(),
+        };
+
+        // Whenever the URL changes, we need to re-generate any short URLs that are displayed.
```
I think this is a better user experience, but I'm worried that it could create a bunch of never-used short urls (which in turn are elasticsearch docs)

spalger commented:
Once you enable sharing you have to close and re-open the share ui to go back to full URLs.

Bargs commented:
Coming to this with semi-fresh eyes, the term Original URL still seems confusing to me. I had to re-read the descriptive text a few times to remember everything we talked about, and users won't have that context. I've been mulling it over and I have a few questions:

* Is it useful for the snapshot URLs to contain a link to the vis they were loaded from? A snapshot is a bundle of state with no real relation to the original saved vis. Perhaps we should fully embrace that fact and remove the vis ID from the URL. That way all snapshots are created equal, whether they originated from a saved vis or not. This makes the behavior more predictable for users, snapshots are always a purely transient entity that allow you to share the state of a vis/dashboard as you currently see it, without saving. It would also allow us to remove this confusing line of text, which makes no sense if the Vis was never saved: "Edits to the saved visualization won't be visible via this URL."
* If you buy into the above what would you think about changing "Share Original" to "Share Saved Viz"/"Share Saved Dashboard" or maybe just "Share Link"? If we do the first bullet, I think that gives us an opportunity to better separate the two sharing mechanisms. One mechanism shares links to named things that are saved (and this becomes the primary way things are shared). The other mechanism shares transient snapshots (this becomes more of a special use case for ad hoc reports).
* We can check the dirty state of the app to see if changes have been made since the last save (like in reporting). Maybe we should dynamically display a warning on the Original URL side if there are unsaved changes.
* Building on all of the above, maybe we should only display one type of sharing at a time, instead of placing them side-by-side? For an unsaved vis, you would only see the snapshot option which would include a message about saving the vis if you want to share future changes. If the vis was saved at some point in the past, we show the "original url" view by default. If we detect unsaved changes, we display a warning message that says you should save your changes or if you'd like to share a snapshot, "click here" and it'll switch to the snapshot view. I think we could even provide the short url by default in this scenario since it's the recommended URL and it would only get generated when a user explicitly says they want a snapshot.

Sorry if this is a giant rambling mess, I've been thinking though it as I type. I guess the overall thing I'm trying to get at is, I think we can create a better separation between the two sharing mechanisms and then guide the user to the correct one based on the current context. Rather than display everything all the time and force the user to read and understand a bunch of stuff right off the bat, we can guide them to the correct decision by hiding inappropriate options and providing helpful, dynamic messages.

cjcenizal commented:
### Removing reference to original visualization from Snapshot URLs

I think your argument makes sense and I think if we want to preserve a reference to the original, we should do that by persisting that reference in a query param, as opposed to part of the URL. This will make it clear that it's simply a reference, and not actually a contributing factor to the way the visualization is rendered. I think we should do this in a subsequent PR though.

## Share Saved vs Share Original

I'm down with "Share Saved"! Good idea!

## Unsaved warning message

Another good idea. I think this is best tackled in another PR.

## Showing different UIs based on different states

I think this is worth exploration, but it's also not a blocker for this PR. Are you OK with merging this after I make the change to the "Original" naming pattern, and then we can explore next steps for improving it?

cjcenizal commented:
Updated UI per @Bargs suggestion:

![image](https://cloud.githubusercontent.com/assets/1238659/18456189/e3573c7e-7903-11e6-995d-f15c8c61f56e.png)
![image](https://cloud.githubusercontent.com/assets/1238659/18456191/e5d470a2-7903-11e6-8cf6-bb795ade160e.png)

epixa commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,146 @@
+      }
+
+      // Default to allowing an embedded IFRAME, unless it's explicitly set to false.
+      this.allowEmbed = $scope.allowEmbed === 'false' ? false : true;
```
How about simply `$scope.allowEmbed !== 'false'` here?  It took me a moment to grok what this what was doing.

spalger replied:
Or combine this and the if statement above into a `switch()`?

cjcenizal replied:
I tried the switch idea, but couldn't figure out a case to match the `$scope.allowEmbed !== undefined` condition.

Re the original suggestion, I think equality comparisons are easier to think about than inequality, especially when comparing booleans. When you see an equality comparison in a ternary, you can associate the happy path with the values and variables in the comparison. But with an inequality comparison, you need to think of the opposite of the compared value to imagine the happy path, which I think is one extra unnecessary step.

Is the ternary the problem? Do you think this is easier to follow @epixa ?

```javascript
this.allowEmbed = true;

// Default to allowing an embedded IFRAME, unless it's explicitly set to false.
if ($scope.allowEmbed === 'false') {
  this.allowEmbed = false;
}
```

I can also make the comment clearer if you have any suggestions.

epixa replied:
I'm not sure if that makes it clearer or not, but I don't think it's worth splitting this into different statements.  Honestly though, I don't feel strongly about it one way or another.

spalger replied:
```
function castAllowEmbedToBoolean(
  switch ($scope.allowEmbed) {
    case 'true':
      return true;
    case undefined:
    case 'false':
      return false;
    default:
      throw new TypeError('...');
  }
}
```

cjcenizal replied:
Thanks @spalger , I see what you're saying. I prefer the way it is currently because I think it's clearer that `true` is the default value _unless_ "false" is passed in.

epixa commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,146 @@
+        const url = $location.absUrl();
+        // Replace hashes with original RISON values.
+        const urlWithStates = unhashUrl(url, getUnhashableStates());
+        return urlWithStates;
```
You can return on the same line here rather than creating a new assignment.

epixa commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,146 @@
+        });
+
+        // Select the text to be copied. If the copy fails, the user can easily copy it manually.
+        const copyTextarea = document.querySelector(selector);
```
Minor nitpick here: but if you use angular's `$document` service instead of directly accessing the global `document`, it makes testing a good deal easier.

cjcenizal replied:
Sounds good but how does it make it easier?

epixa replied:
You can mock the service so it doesn't actually require dom access and instead verify that the appropriate selector is being passed.

Bargs commented:
Embedded iframe isn't an option for Saved object sharing?

Bargs commented:
<img width="1010" alt="screen shot 2016-09-12 at 8 13 45 pm" src="https://cloud.githubusercontent.com/assets/6239176/18457303/7f996492-7925-11e6-84c4-26954de2bd9d.png">

We're mixing different title casing styles here, should it be "Share Saved Dashboard" or "Share snapshot"?

cjcenizal commented:
Do we have copy-writing rules in place? Personally I prefer sentence-casing (capitalizing the first word) over title-casing (capitalizing nouns and verbs) in general, and using capital words to indicate special terms, like "Snapshot".

Bargs commented:
I was sort of thinking of "Saved Dashboard" as a special term, in contrast to "Snapshot". Not sure about copy rules though. @debadair ?

Bargs commented:
Other than the two things I mentioned above, functionality LGTM

cjcenizal commented:
@Bargs I added the option to embed an iframe to a saved visualization, but I added on the global state to the URL, so that it won't just load the default time range. Let's fix the capitalization later, after @debadair chimes in.

epixa commented:
If you rebase this, the build should pass.

spalger commented:
LGTM

spalger commented:
Merged locally so I could run it

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -39,15 +41,16 @@ app.directive('share', function (Private) {
         // Remove the query string.
-        const url = $location.absUrl();
-        return url.split('?')[0];
+        const baseUrl = $location.absUrl().split('?')[0];
```
This really should be using bonafide url parsing

spalger commented on src/ui/public/share/directives/share.js:
```diff
@@ -1,18 +1,149 @@
+        const baseUrl = $location.absUrl().split('?')[0];
+
+        // Add global state to the URL so that the iframe doesn't just show the time range default.
+        return `${baseUrl}?_g=${$location.search()._g}`;
```
This really needs to use url parsing and formatting

cjcenizal replied:
How's this:

```javascript
const parsedUrl = parseUrl(window.location.href);
return formatUrl({
  protocol: parsedUrl.protocol,
  auth: parsedUrl.auth,
  host: parsedUrl.host,
  path: parsedUrl.path,
  search: {
    _g: parsedUrl.search._g,
  },
});
```

I can definitely see how this is more consistent with the rest of the codebase. Are there other reasons to write it this way?

spalger replied:
You actually need to parse the parent url and the url in the hash, beyond that I'm pretty sure that `search` needs to be `query` (search is for strings, query for objects, iirc).

As to why, adding a query string param to the non-hash part of the url would completely break this functionality:

```
https://localhost:5601/cdf/app/kibana?flag=true#/discover/New-Saved-Search?_g=()&_a=(columns:!(_source),filters:!(),index:'logstash-*',interval:auto,query:(query_string:(analyze_wildcard:!t,query:'*')),sort:!('@timestamp',desc))
```
produces the share url
```
https://localhost:5601/cdf/app/kibana?_g=()
```

When it comes to parsing standard formats though, I always urge people to use the proper encoding/decoding tools. While simply splitting on '?' works today, there is good reason to expect that it at some point an edge case will show up that may or may-not be very hard to debug.

cjcenizal replied:
Oh, I see what you're saying! I forgot about that gotcha about the way Angular stores stuff after the hash. Great call, thanks.

spalger commented:
LGTM

epixa commented:
@cjcenizal Can you make sure the docs are updated with new screenshots and such? https://www.elastic.co/guide/en/kibana/current/dashboard.html#sharing-dashboards

That doesn't necessarily have to happen in this PR if you're still trying to get this feature in for 5.0, but it should happen in a follow up as soon as possible.

epixa commented:
@cjcenizal I created backport PRs for this since we're running out of time to get stuff into 5.0 and I wanted the builds going.

cjcenizal commented:
Thanks @epixa
